File: app/src/main/java/com/stathis/elmepaunivapp/ui/department/recyclerviews/DepMembersAdapter.java
package com.stathis.elmepaunivapp.ui.department.recyclerviews;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.DiffItemCallbackClass;
import com.stathis.elmepaunivapp.listeners.DepMembersClickListener;
import com.stathis.elmepaunivapp.models.DepMembers;

public class DepMembersAdapter extends ListAdapter<DepMembers,DepMembersViewHolder> {

    private DepMembersClickListener listener;

    public DepMembersAdapter(DepMembersClickListener listener) {
        super(new DiffItemCallbackClass<DepMembers>());
        this.listener = listener;
    }

    @NonNull
    @Override
    public DepMembersViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.dep_members_item_row, parent, false);
        return new DepMembersViewHolder(view, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull DepMembersViewHolder holder, int position) {
        holder.present(getItem(position));
    }
}

File: feature/src/main/java/com/stathis/feature/ui/department/adapter/DepMembersAdapter.kt
package com.stathis.feature.ui.department.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.ListAdapter
import com.stathis.core.base.BaseDiffUtil
import com.stathis.core.base.BaseViewHolder
import com.stathis.model.UiModel
import com.stathis.feature.BR
import com.stathis.feature.R
import com.stathis.feature.databinding.HolderDepmemberItemBinding
import com.stathis.feature.databinding.HolderEmptyViewBinding
import com.stathis.feature.databinding.HolderShimmerDepmemberItemBinding
import com.stathis.model.general.ShimmerItem
import com.stathis.model.personnel.Person

class DepMembersAdapter : ListAdapter<UiModel, DepMembersViewHolder>(BaseDiffUtil<UiModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DepMembersViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = when (viewType) {
            R.layout.holder_depmember_item -> {
                HolderDepmemberItemBinding.inflate(inflater, parent, false)
            }

            R.layout.holder_shimmer_depmember_item -> {
                HolderShimmerDepmemberItemBinding.inflate(inflater, parent, false)
            }

            else -> HolderEmptyViewBinding.inflate(inflater, parent, false)
        }
        return DepMembersViewHolder(view)
    }

    override fun onBindViewHolder(holder: DepMembersViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun getItemViewType(position: Int) = when (getItem(position)) {
        is Person -> R.layout.holder_depmember_item
        is ShimmerItem -> R.layout.holder_shimmer_depmember_item
        else -> R.layout.holder_empty_view
    }
}

class DepMembersViewHolder(
    private val binding: ViewDataBinding
) : BaseViewHolder(binding) {

    override fun bind(data: UiModel) {
        binding.setVariable(BR.model, data)
    }
}